### PR TITLE
global-state-update-gen: don't return validator set if it did not change

### LINKED ISSUE #3518: Change global-state-update-gen tool to not always add validators in the output file
The tool currently always adds a validators entry to the output global_state.toml file.  It should now only add this section if the set of validators is being changed in the upgrade.  The GlobalStateUpdateConfig and GlobalStateUpdate structs should have their validators fields made optional.

It may be worth to go through the NCTL test that use the tool and remove the soon gonna-be unnecessary calls to change_validators . This could be one such place, for example: https://github.com/casper-network/casper-node/blob/feat-fast-sync-v2/utils/nctl/sh/assets/setup_shared.sh#L470

### PR DESCRIPTION
**Pull request by alsrdn**
Don't return validator set if it did not change when upgrading.
Include validator weights in the output of the validator set.
Also change parsing of `global_state.toml` to reflect these changes.

Signed-off-by: Alexandru Sardan <alexandru@casperlabs.io>

Fixes: https://github.com/casper-network/casper-node/issues/3518

### CODE REVIEW

File: utils/global-state-update-gen/src/generic.rs [hunk 1]
```diff
@@ -129,16 +130,25 @@ fn update_auction_state<T: StateReader>(
...
 
-    // All entries in the new snapshot contain the same set of validators, just use the first entry.
-    new_snapshot
-        .values()
-        .next()
-        .expect("snapshot should have at least one entry")
-        .keys()
-        .cloned()
-        .collect()
+        if validators_diff.added.is_empty() && validators_diff.removed.is_empty() {
```
Comment (fizyk20) [inline utils/global-state-update-gen/src/generic.rs]: I don't think that's the correct condition. It's possible to twiddle the stakes of the validators without adding or removing any of them - and in such a case that should still be `Some` (especially when we add the weights to the picture).
I think this should say something along the lines of `accounts.iter().all(|account| account.validator.is_none())`.

File: utils/global-state-update-gen/src/generic.rs [hunk 2]
```diff
@@ -129,16 +130,25 @@ fn update_auction_state<T: StateReader>(
...
+            None
+        } else {
+            // All entries in the new snapshot contain the same set of validators,
+            // just use the first entry.
```
Comment (fizyk20) [inline utils/global-state-update-gen/src/generic.rs]: Not added in this PR, but this comment might actually be incorrect in some cases :grimacing: Feel free to leave it as is for now, but if you want to change it, I suggest something like:

> // We need to output the validators for the next era, which are contained in the first entry in the snapshot.

File: utils/global-state-update-gen/src/generic.rs (hunk 1 above)
Comment (alsrdn) [reply utils/global-state-update-gen/src/generic.rs]: Yes, that's correct. I'll add a new commit which also adds weights and will change this condition as well.

**Review by fizyk20 — APPROVED** [review]

**Review by Fraser999 — APPROVED** [review]

**Comment by alsrdn** [comment]
bors merge

**Pull request merged**